File: maxStack.js
// Implement the push, pop, and getMax methods

class MaxStack {
  constructor() {
    this.stack = new Stack();
    this.maximumsStack = new Stack();
  }

  push(item) {
    this.stack.push(item);
    if (this.maximumsStack.peek() === null || item >= this.maximumsStack.peek()) {
        this.maximumsStack.push(item);
    }
  }

  pop() {
    let item = this.stack.pop();
    if (item === this.maximumsStack.peek()) {
      this.maximumsStack.pop();
    }
    return item; 
  }

  getMax() {
    return this.maximumsStack.peek(); 
  }
}

class Stack {
  constructor() {

    // Initialize an empty stack
    this.items = [];
  }

  // Push a new item onto the stack
  push(item) {
    this.items.push(item);
  }

  // Remove and return the last item
  pop() {

    // If the stack is empty, return null
    // (It would also be reasonable to throw an exception)
    if (!this.items.length) {
      return null;
    }
    return this.items.pop();
  }

  // Return the last item without removing it
  peek() {
    if (!this.items.length) {
      return null;
    }
    return this.items[this.items.length - 1];
  }
}




/*

O- integer representing largest element in a stack
I- "this" representing the stack 
C- stack only holds integers (can be negative)
E- nothing in stack, same size items


approach

functions that: add an item, remove an item, look at the last item
can we mutate original input stack? 
do we have to care about increase in size complexity 


*/


// Tests

// const s = new MaxStack();
// s.push(5);

// assertEquals(5, s.getMax(), 'check max after 1st push');

// s.push(4);
// s.push(7);
// s.push(7);
// s.push(8);

// assertEquals(8, s.getMax(), 'check before 1st pop');
// assertEquals(8, s.pop(), 'check pop #1');
// assertEquals(7, s.getMax(), 'check max after 1st pop');
// assertEquals(7, s.pop(), 'check pop #2');
// assertEquals(7, s.getMax(), 'check max after 2nd pop');
// assertEquals(7, s.pop(), 'check pop #3');
// assertEquals(5, s.getMax(), 'check max after 3rd pop');
// assertEquals(4, s.pop(), 'check pop #4');
// assertEquals(5, s.getMax(), 'check max after 4th pop');

// function assertEquals(a, b, desc) {
//   if (a === b) {
//     console.log(`${desc} ... PASS`);
//   } else {
//     console.log(`${desc} ... FAIL: ${a} != ${b}`);
//   }
// }

File: maxStack.py
import unittest


class Stack(object):

    def __init__(self):
        """Initialize an empty stack"""
        self.items = []

    def push(self, item):
        """Push a new item onto the stack"""
        self.items.append(item)

    def pop(self):
        """Remove and return the last item"""
        # If the stack is empty, return None
        # (it would also be reasonable to throw an exception)
        if not self.items:
            return None

        return self.items.pop()

    def peek(self):
        """Return the last item without removing it"""
        if not self.items:
            return None
        return self.items[-1]


class MaxStack(object):

    # Implement the push, pop, and get_max methods

    def __init__(self):
        self.stack = Stack()
        self.maxes_stack = Stack()

    def push(self, item):
        self.stack.push(item)
        if item >= self.maxes_stack.peek() or self.maxes_stack.peek == None:
            self.maxes_stack.push(item)

    def pop(self):
        if self.stack.peek() == self.maxes_stack.peek():
            self.maxes_stack.pop()
        return self.stack.pop()

    # Ouput: the largest element in the stack
    # Input: none
    # Constraints: assume stack only holds integers
    # Edge cases: nothing on stack, multiple same size
    def get_max(self):
        return self.maxes_stack.peek()

#
# # Tests
#
# class Test(unittest.TestCase):
#
#     def test_stack_usage(self):
#         max_stack = MaxStack()
#
#         max_stack.push(5)
#
#         actual = max_stack.get_max()
#         expected = 5
#         self.assertEqual(actual, expected)
#
#         max_stack.push(4)
#         max_stack.push(7)
#         max_stack.push(7)
#         max_stack.push(8)
#
#         actual = max_stack.get_max()
#         expected = 8
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.pop()
#         expected = 8
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.get_max()
#         expected = 7
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.pop()
#         expected = 7
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.get_max()
#         expected = 7
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.pop()
#         expected = 7
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.get_max()
#         expected = 5
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.pop()
#         expected = 4
#         self.assertEqual(actual, expected)
#
#         actual = max_stack.get_max()
#         expected = 5
#         self.assertEqual(actual, expected)
#
#
# unittest.main(verbosity=2)